framework/src/utils/FillBetweenPointVectorsTools.C

```diff
@@ -292,56 +309,117 @@ elementsCreationFromNodesVectors(ReplicatedMesh & mesh,
+          elem->set_node(1) = nodes[i][nodes_down_it + 1];
+          is_elem_flip = true;
+        }
         elem->subdomain_id() = transition_layer_id;
         if (i == 0)
           boundary_info.add_side(elem, 1, input_boundary_1_id);
         if (nodes_up_it == 0 && nodes_down_it == 0)
-          boundary_info.add_side(elem, 0, begin_side_boundary_id);
+          boundary_info.add_side(elem, is_elem_flip ? 2 : 0, begin_side_boundary_id);
         nodes_down_it++;
       }
     }
     // Handle the end
     while (nodes_up_it < node_number_up - 1)
     {
       Elem * elem = mesh.add_elem(new Tri3);
-      elem->set_node(0) = nodes[i + 1][nodes_up_it];
-      elem->set_node(1) = nodes[i][nodes_down_it];
-      elem->set_node(2) = nodes[i + 1][nodes_up_it + 1];
+      // Adjust the order of nodes in an element so that the mesh can be extruded in (0 0 1)
+      // direction.
+      bool is_elem_flip(false);
+      if (((*nodes[i][nodes_down_it] - *nodes[i + 1][nodes_up_it])
+               .cross((*nodes[i + 1][nodes_up_it + 1] - *nodes[i + 1][nodes_up_it]))
+               .unit())(2) > 0)
+      {
+        elem->set_node(0) = nodes[i + 1][nodes_up_it];
+        elem->set_node(1) = nodes[i][nodes_down_it];
+        elem->set_node(2) = nodes[i + 1][nodes_up_it + 1];
+      }
+      else
+      {
+        elem->set_node(0) = nodes[i + 1][nodes_up_it];
+        elem->set_node(2) = nodes[i][nodes_down_it];
+        elem->set_node(1) = nodes[i + 1][nodes_up_it + 1];
+        is_elem_flip = true;
+      }
       elem->subdomain_id() = transition_layer_id;
       nodes_up_it++;
       if (i == num_layers - 1)
-        boundary_info.add_side(elem, 2, input_boundary_2_id);
+        boundary_info.add_side(elem, is_elem_flip ? 0 : 2, input_boundary_2_id);
       if (nodes_up_it == node_number_up - 1 && nodes_down_it == node_number_down - 1)
         boundary_info.add_side(elem, 1, end_side_boundary_id);
     }
     while (nodes_down_it < node_number_down - 1)
     {
       Elem * elem = mesh.add_elem(new Tri3);
-      elem->set_node(0) = nodes[i + 1][nodes_up_it];
-      elem->set_node(1) = nodes[i][nodes_down_it];
-      elem->set_node(2) = nodes[i][nodes_down_it + 1];
+      // Adjust the order of nodes in an element so that the mesh can be extruded in (0 0 1)
+      // direction.
+      bool is_elem_flip(false);
+      if (((*nodes[i][nodes_down_it] - *nodes[i + 1][nodes_up_it])
+               .cross((*nodes[i][nodes_down_it + 1] - *nodes[i + 1][nodes_up_it]))
+               .unit())(2) > 0)
+      {
+        elem->set_node(0) = nodes[i + 1][nodes_up_it];
+        elem->set_node(1) = nodes[i][nodes_down_it];
+        elem->set_node(2) = nodes[i][nodes_down_it + 1];
+      }
+      else
+      {
+        elem->set_node(0) = nodes[i + 1][nodes_up_it];
+        elem->set_node(2) = nodes[i][nodes_down_it];
+        elem->set_node(1) = nodes[i][nodes_down_it + 1];
+        is_elem_flip = true;
+      }
```

Comment:
since you have 3 times the same code we could have a routine in the tools?

Reply:
Yes, you are right